service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationQueries.kt

```diff
@@ -542,3 +546,224 @@ fun Database.Read.getReservationContractDayRanges(
+                       AND ar.date = :examinationDate
+         LEFT JOIN daily_service_time dst
+                   ON pcd.child_id = dst.child_id
+                       AND dst.validity_period @> :examinationDate
+            """
+                .trimIndent()
+        )
+        .bind("unitId", unitId)
+        .bind("examinationDate", day)
+        .toList<DailyChildReservationInfoRow>()
+}
+
+data class GroupReservationStatisticsRow(
+    val date: LocalDate,
+    val groupId: GroupId?,
+    val calculatedPresent: BigDecimal,
+    val absent: Int,
+    val present: Int
+)
+
+fun Database.Read.getReservationStatisticsForUnit(
+    confirmedDays: List<LocalDate>,
+    unitId: DaycareId
+): Map<LocalDate, List<GroupReservationStatisticsRow>> {
+    return createQuery(
+            """
+select count(1) filter (where a.absence_status = 'ABSENT')     as absent,
+       count(1) filter (where a.absence_status = 'PRESENT') as present,
+       coalesce(sum(occupancy_coefficient) filter ( where a.absence_status = 'PRESENT' ), 0)              as calculated_present,
+       date,
+       group_id
+from (with placement_days as
+               (SELECT day.date,
+                       pl.child_id,
+                       pl.start_date        as placement_start,
+                       pl.end_date          as placement_end,
+                       gpl.start_date       as grouping_start,
+                       gpl.end_date         as grouping_end,
+                       gpl.daycare_group_id as group_id
+                FROM (SELECT t.date
+                      FROM unnest(:confirmedDays) AS t(date)) day
+                         LEFT JOIN placement pl
+                                   ON daterange(pl.start_date, pl.end_date, '[]') @> day.date AND pl.unit_id = :unitId
+                         JOIN daycare_group_placement gpl
+                                   ON pl.id = gpl.daycare_placement_id AND
+                                      daterange(gpl.start_date, gpl.end_date, '[]') @> day.date
+                WHERE NOT EXISTS (select
+                                  from backup_care bc
+                                  where bc.child_id = pl.child_id
+                                    AND bc.unit_id = pl.unit_id
+                                    AND daterange(bc.start_date, bc.end_date, '[]') @> day.date)
+                UNION
+                SELECT day.date,
+                       bc.child_id,
+                       bc.start_date as placement_start,
+                       bc.end_date   as placement_end,
+                       bc.start_date as grouping_start,
+                       bc.end_date   as grouping_end,
+                       bc.group_id   as group_id
+                FROM (SELECT t.date
+                      FROM unnest(:confirmedDays) AS t(date)) day
+                         LEFT JOIN backup_care bc ON daterange(bc.start_date, bc.end_date, '[]') @> day.date
+                    AND bc.unit_id = :unitId)
```

Comment:
This part could probably also use the `realized_placement_one` function. It it cannot be used because of performance problems, this query should at least take some ideas from there, because in the current form it's quite unreadable.

Reply:
I refactored the query to utilize `realized_placement_one` and it simplified the query strucure significantly, although it also introduced new feature specific incoming/outgoing backup care deduction logic. The addition of term breaks also added some logic (and duplicated logic at that), but the end result is still more readable than the original.

I tried to figure a way of doing the statistics gathering on Kotlin side but apart from the duplicated logic it seemed even more complicated than the group by query.